lxd/devlxd.go

```diff
@@ -103,7 +103,13 @@ var devlxdImageExport = devLxdHandler{"/1.0/images/{fingerprint}/export", func(d
 
 var devlxdMetadataGet = devLxdHandler{"/1.0/meta-data", func(d *Daemon, c instance.Instance, w http.ResponseWriter, r *http.Request) *devLxdResponse {
 	value := c.ExpandedConfig()["user.meta-data"]
-	return okResponse(fmt.Sprintf("#cloud-config\ninstance-id: %s\nlocal-hostname: %s\n%s", c.Name(), c.Name(), value), "raw")
+
+	id := c.LocalConfig()["volatile.cloud-init.instance-id"]
+	if id == "" {
+		id = c.Name()
```

Comment:
@stgraber why do we do this out of interest?

Reply:
To handle upgrades. For existing instances, we don't want the `instance-id` to change on upgrade when not combined with a change to a key that triggers a new instance-id.

If we were to generate an instance-id UUID on upgrade, it would cause cloud-init to re-trigger in all instances potentially re-generating system configs (like netplan) and breaking instances.

Reply:
This is also why we don't generate it when missing in startCommon as doing so would cause all insances to change instance-id.